optimize/cmaes.go

```diff
@@ -0,0 +1,381 @@
+// is iterated until convergence.
+//
+// As the normal distribution is progressively updated according to the best samples,
+// it can be that the mean of the distribution is updated in a gradient-descent
+// like fashion, followed by a shrinking covariance.
+// It is recommended that the algorithm be run multiple times (with different
+// InitMean) to have a better chance of finding the global minimum.
+//
+// The CMA-ES-Chol algorithm differs from the standard CMA-ES algorithm in that
+// it directly updates the Cholesky decomposition of the normal distribution.
+// This changes the runtime is O(dim^2*pop) rather than O(dim^3). The evolution
+// of the multi-variate normal will be similar to the baseline CMA-ES algorithm,
+// but the covariance update equation is not identical.
+//
+// For more information about the CMA-ES algorithm, see
+//  https://en.wikipedia.org/wiki/CMA-ES
+//  https://arxiv.org/pdf/1604.00772.pdf
+type CmaEsChol struct {
+	// InitStepSize sets the initial size of the covariance matrix adaptation.
+	// If InitStepSize is 0, a default vaule of 0.5 is used. InitStepSize cannot
+	// be negative.
+	InitStepSize float64
+	// Population sets the population size for the algorithm. If Population is
+	// 0, a default value of 4 + math.Floor(3*math.Log(float64(dim))) is used.
+	// Population cannot be negative.
+	Population int
+	// InitMean is the initial mean of the multivariate normal for sampling
+	// input locations. If InitMean is nil, the zero vector is used. If InitMean
+	// is not nil, it must have length equal to the problem dimension.
+	InitMean []float64
+	// InitCholesky specifies the Cholesky decomposition of the covariance
+	// matrix for the initial sampling distribution. If InitCholesky is nil,
+	// a default value of I is used. If it is non-nil, then it must have
+	// InitCholesky.Size() be equal to the problem dimension.
+	InitCholesky *mat.Cholesky
+	// StopLogDet sets the threshold for stopping the optimization if the
+	// distribution becomes too peaked. The log determinant is a measure of the
+	// (log) "volume" of the normal distribution, and when it is too small
+	// the samples are almost the same. If the log determinant of the covariance
+	// matrix becomes less than StopLogDet, the optimization run is concluded.
+	// If StopDeterminant is 0, a default value of dim*log(1e-16) is used.
+	// If StopDeterminant is NaN, the stopping criteria is not used, though
+	// this can cause numeric instabilities in the algorithm.
+	StopLogDet float64
+	// ForgetBest, when true, does not track the best overall function value found,
+	// instead returning the new best sample in each iteration. If ForgetBest
+	// is false, then the minimum value returned will be the lowest across all
+	// iterations, regardless of when that sample was generated.
+	ForgetBest bool
+	// Src allows a random number generator to be supplied for generating samples.
+	// If Src is nil the generator in math/rand is used.
+	Src *rand.Rand
```

Comment:
I don't remember the outcome of our previous discussion: do we call `rand.Rand` a `Src` or a `Source` ?

Reply:
It's Src pretty much everywhere (and Source a few places)

Reply:
ok.

indeed:
```
$> git grep "Source" -- *.go | grep rand.Rand
stat/distuv/bernoulli.go:17:	Source *rand.Rand
stat/distuv/beta.go:29:	Source *rand.Rand
stat/distuv/exponential.go:18:	Source *rand.Rand
stat/distuv/f.go:25:	Source *rand.Rand
stat/distuv/gamma.go:29:	Source *rand.Rand
stat/distuv/laplace.go:20:	Source *rand.Rand
stat/distuv/lognormal.go:18:	Source *rand.Rand
stat/distuv/norm.go:23:	Source *rand.Rand
stat/distuv/triangle.go:15:	Source  *rand.Rand
stat/distuv/uniform.go:20:	Source *rand.Rand
stat/distuv/weibull.go:22:	Source *rand.Rand

$> git grep "Src" -- *.go | grep rand.Rand | wc -l
9

$> git grep "Source" -- *.go | grep rand.Rand | wc -l
11
```

Reply:
It's an issue to consider the change. I've had no bites there though.